stew/templateutils.nim

```diff
@@ -0,0 +1,53 @@
+type CppVar[T] = distinct ptr T
+
+iterator evalTemplateParamOnceImpl[T](x: T): lent T =
+  yield x
+
+when defined(cpp):
+  # TODO `nim cpp` miscompiles iterators returning `var`,
+  #      so we need to emulate them in terms of pointers:
+  iterator evalTemplateParamOnceImpl[T](x: var T): CppVar[T] =
+    yield CppVar[T](addr(x))
+
+  template stripCppVar[T](p: CppVar[T]): var T =
+    ((ptr T)(p))[]
+else:
+  iterator evalTemplateParamOnceImpl[T](x: var T): var T =
+    yield x
+
+template evalTemplateParamOnce*(templateParam, newName, blk: untyped) =
+  ## This can be used in templates to avoid the problem of multiple
+  ## evaluation of template parameters. Compared to the naive approach
+  ## of introducing an additional local variable, it has two benefits:
+  ##
+  ## * It avoids copying whenever possible.
+  ## * It works for var parameters.
+  ##
+  ## Usage example:
+  ##
+  ## template foo(xParam: SomeType) =
+  ##   evalTemplateParamOnce(xParam, x):
+  ##     echo x
+  ##     echo x
+  ##
+  ##  A currently existing limitation is that the `evalTemplateParamOnce`
```

Comment:
I've found that in most cases this can be worked around by having the template return `untyped` - wouldn't that work here?

ie `template foo(): untyped = "hello"` and `template foo(): untyped = discard` both work

Reply:
No, the limitation is coming from the fact that the body of the block provided by the user is inserted inside a for loop. It's possible to work-around the limitation by introducing a copy, but since the user code can already do that at the call-site of the template, it would be inappropriate to do in within the template.

Reply:
so by and large, this construct sometimes introduces a copy as seen in my example above, is difficult to spell, has complex limitations that are difficult to reason about, and doesn't work with expressions, which is a proven way to make user code safer and easier to reason about in many situations - it doesn't really look like a net win in any case, ie can you proved an example where it does _not_ introduce the copy?

Reply:
Sorry, I've missed your previous comment. I've added a reply now that demonstrates that no unwanted copies are produced in the provided example. You can also look at the provided test cases which further prove that.

Reply:
I will also mention that the `keepItif` operation added to the `sets` module is a good example for something enabled by this API. You cannot emulate it with the usual approach of introducing a local variable. There is also another safe way to implement it by introducing a closure, but this would result in much more surprising codegen side-effects and limitations IMO.